File: Exercícios FPC I/ex12_Canos2.cpp
def maximo(a,b):
  #retorna o maior entre dois inteiros
  if a > b:
    return a
  else:
    return b

def cortarCanosAtualizado(i,t):
  if i == n:
    return 0

  elif t == 0:
    return 0

  elif t < 0:
    return -1

  ans = matriz[i][t]

  if ans == -1:
    ans = maximo(cortarCanosAtualizado(i + 1, t), valor[i] + cortarCanosAtualizado(i, t - peso[i]));
    print(ans)
  return ans;
  
#==============================    
n,t = [int(i) for i in input().split()]

peso = []
valor = []
for i in range(0,n):
    p, v = [int(i) for i in input().split()]
    peso.append(p)
    valor.append(v)

matriz = []

matriz = [[-1 for x in range(t+1)] for x in range(n+1)]


print(cortarCanosAtualizado(0, t))

File: Exercícios FPC I/ex12_Canos2.py
def maximo(a,b):
  #retorna o maior entre dois inteiros
  if a > b:
    return a
  else:
    return b

def cortarCanosAtualizado(n,t):
  # add mais um espaço a lista para ter a parada/ponto de inicio
  # E pq tbm tenho lucro zero sem vender nada
  list_lucros = [0 for x in range(t+1)] 
  for i in range(0,t+1):
    for j in range(0,n):
      if(comprimento[j] <= i):
        list_lucros[i] = max(list_lucros[i - comprimento[j]] + valor[j],list_lucros[i]);
        
  return(list_lucros[t])

#====================================================================   
n,t = [int(i) for i in input().split()]

comprimento = []
valor = []
for i in range(0,n):
    p, v = [int(i) for i in input().split()]
    comprimento.append(p)
    valor.append(v)